### PR TITLE
Feature pull client data from object store

### PR DESCRIPTION
**Pull request by hardbyte**
_Builds on https://github.com/data61/anonlink-entity-service/pull/548_

Adds the REST api changes and backend changes to pull binary encoding data from an object store.

### CODE REVIEW

File: deployment/entity-service/requirements.yaml [hunk 1]
```diff
@@ -8,6 +8,6 @@ dependencies:
     repository: https://kubernetes-charts.storage.googleapis.com
     condition: provision.minio
   - name: postgresql
-    version: 8.3.0
-    repository: https://kubernetes-charts.storage.googleapis.com
+    version: 8.9.1
```
Comment (JoyceXinyueWang) [inline deployment/entity-service/requirements.yaml:11]: How often would you recommend us to check the version of dependent charts? Would there be similar dependency bot for charts?

File: deployment/entity-service/templates/NOTES.txt [hunk 2]
```diff
@@ -1,37 +1,48 @@
...
 
-Soon you should be able to visit the entity service api.
+The Anonlink Entity Service can be accessed via port {{ .Values.api.service.servicePort }} on the
+following DNS name from within your cluster:
+{{ template "api.fullname" . }}.{{ .Release.Namespace }}.svc.cluster.local
 
 1. Get the entity service URL by running:
 
-{{- if contains "NodePort" .Values.api.service.type }}
+{{- if eq .Values.api.service.type "NodePort" "ClusterIP" }}
```
Comment (JoyceXinyueWang) [inline deployment/entity-service/templates/NOTES.txt:10]: Why do we wrap code in double curly bracket? Is it some Javascript?

File: deployment/entity-service/templates/NOTES.txt [hunk 3]
```diff
@@ -1,37 +1,48 @@
...
-  kubectl port-forward SVC_NAME 8080:80
+You may wish to connect to the API using anonlink-client. https://github.com/data61/anonlink-client/
+The client can be installed using pip:
 
-And visit http://127.0.0.1:8080/
+    $ pip install anonlinkclient
+    $ anonlink status --server=<SERVER>
 
-{{- end }}
+Issues with the Anonlink Entity Service can be created on Github - https://github.com/data61/anonlink-entity-service/
```
Comment (JoyceXinyueWang) [inline deployment/entity-service/templates/NOTES.txt:48]: Thanks for the session this morning, these notes finally make some sense to me :P

File: backend/entityservice/encoding_storage.py [hunk 4]
```diff
@@ -110,3 +121,54 @@ def get_encoding_chunk(conn, chunk_info, encoding_size=128):
...
+
+def upload_clk_data_binary(project_id, dp_id, encoding_iter, receipt_token, count, size=128):
+    """
+    Save the user provided binary-packed CLK data.
+
+    """
+    filename = None
+    # Set the state to 'pending' in the uploads table
+    with DBConn() as conn:
+        db.insert_encoding_metadata(conn, filename, dp_id, receipt_token, encoding_count=count, block_count=1)
```
Comment (wilko77) [inline backend/entityservice/encoding_storage.py:133]: what's with the `filename = None` business?

File: backend/entityservice/encoding_storage.py [hunk 5]
```diff
@@ -110,3 +121,54 @@ def get_encoding_chunk(conn, chunk_info, encoding_size=128):
...
+        with opentracing.tracer.start_span('upload-encodings-to-db', child_of=parent_span):
+            store_encodings_in_db(conn, dp_id, encoding_iter, size)
+
+        with opentracing.tracer.start_span('update-encoding-metadata', child_of=parent_span):
+            db.update_encoding_metadata(conn, filename, dp_id, 'ready')
+
+    # # Now work out if all parties have added their data
+    # if clks_uploaded_to_project(project_id):
+    #     logger.info("All parties data present. Scheduling any queued runs")
+    #     check_for_executable_runs.delay(project_id, serialize_span(parent_span))
```
Comment (wilko77) [inline backend/entityservice/encoding_storage.py:158]: where is `check_for_executable_runs` called in the case of external data?

**Comment by hardbyte** [comment]
> 
> 
> The goal for this data pull from object store feature is that we can support population sized linkages. Without blocking, this is not really achievable. After all the hard work you put in we are so close now to achieve this. Why did you decide to not handle blocking information as well?

I hope this wasn't a surprise given it was mentioned in the [design doc](https://github.com/data61/anonlink-entity-service/blob/develop/docs/designs/anonlink-upload-data-plan.md#modification-to-the-upload-endpoint). I specified that only encodings would be supported in the first implementation due to the time available. I wasn't sure if we'd have to create a new binary format for blocking info and that didn't seem too related to the core functionality of "pulling data". However, I have developed a branch that supports external blocking information (see https://github.com/data61/anonlink-entity-service/pull/551).

File: azure-pipelines.yml [hunk 6]
```diff
@@ -298,14 +298,22 @@ stages:
...
             api.ingress.enabled:false
             workers.replicaCount:4
             api.www.image.repository:$(frontendImageName)
+            api.www.image.pullPolicy:Always
             api.app.image.repository:$(backendImageName)
+            api.app.image.pullPolicy:Always
             api.dbinit.image.repository:$(backendImageName)
             workers.image.repository:$(backendImageName)
+            workers.image.pullPolicy:Always
+            anonlink.objectstore.secure:false
```
[the remark this answers is not in the document]
Comment (hardbyte) [reply azure-pipelines.yml:311]: These overrides are translated into command line arguments to `helm template` much like setting individual values when deploying a chart (e.g., `--set anonlink.objectstore.secure=false`). The path in question `anonlink.objectstore` is defined in the `values.yaml` file.

You can see the final command [here](https://dev.azure.com/data61/Anonlink/_build/results?buildId=1971&view=logs&j=9e19c0c0-8b45-571d-b5f1-1c7371127d08&t=8aa1c3fe-e654-5284-604c-8738a640114a):

```
/opt/hostedtoolcache/helm/2.16.3/x64/linux-amd64/helm template /home/vsts/work/1/s/deployment/entity-service --name azk09d6034 --namespace test-azure --set global.postgresql.postgresqlPassword=notaproductionpassword --set api.ingress.enabled=false --set workers.replicaCount=4 --set api.www.image.repository=data61/anonlink-nginx --set api.www.image.pullPolicy=Always --set api.app.image.repository=data61/anonlink-app --set api.app.image.pullPolicy=Always --set api.dbinit.image.repository=data61/anonlink-app --set workers.image.repository=data61/anonlink-app --set workers.image.pullPolicy=Always --set anonlink.objectstore.secure=false --set anonlink.objectstore.uploadEnabled=true --set anonlink.objectstore.uploadAccessKey=testUploadAccessKey --set anonlink.objectstore.uploadSecretKey=testUploadSecretKey --set minio.accessKey=testMinioAccessKey --set minio.secretKey=testMinioSecretKey
```

File: backend/entityservice/api_def/openapi.yaml [hunk 7]
```diff
@@ -1077,17 +1087,92 @@ components:
...
+            - $ref: '#/components/schemas/EncodingArray'
+            - $ref: '#/components/schemas/ExternalData'
+        blocks:
+          oneOf:
+            - $ref: '#/components/schemas/BlockMap'
+            ## TODO may be useful to handle external blocking data too
+            #- $ref: '#/ExternalData'
+
+    EncodingArray:
+      description: Array of encodings, comprising the base64-encoding (and encoding id?).
```
[the remark this answers is not in the document]
Comment (hardbyte) [reply backend/entityservice/api_def/openapi.yaml:1106]: Eek, yeah I wasn't sure about taking the extra step at this point to allow the user to provide the encoding ids. I decided it wasn't part of this feature. I'll remove the question from the api spec

File: backend/entityservice/api_def/openapi.yaml [hunk 8]
```diff
@@ -1077,17 +1087,92 @@ components:
...
+      items:
+        anyOf:
+          - type: string
+            format: byte
+            description: Base64 encoded CLK data
+          - type: string
+          - type: integer
+
+
+    BlockMap:
```
[the remark this answers is not in the document]
Comment (hardbyte) [reply backend/entityservice/api_def/openapi.yaml:1117]: No it isn't required, see above where it is referenced - only the `encodings` are required:

> 
```
EncodingUpload:
      description: Object that contains one data provider's encodings
      type: object
      required: [encodings]
      properties:
        encodings:
          ...
        blocks:
          oneOf:
            - $ref: '#/components/schemas/BlockMap'
```

File: deployment/entity-service/requirements.yaml (hunk 1 above)
Comment (hardbyte) [reply deployment/entity-service/requirements.yaml:11]: I would do it anytime you change the kubernetes deployment template.

File: deployment/entity-service/templates/NOTES.txt (hunk 2 above)
Comment (hardbyte) [reply deployment/entity-service/templates/NOTES.txt:10]: No it is helm template specific - a combination of the Go template language with some extra functions and helm deployment data. See the docs [here](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/)

File: backend/entityservice/encoding_storage.py (hunk 4 above)
Comment (hardbyte) [reply backend/entityservice/encoding_storage.py:133]: It is just a way of keeping the column in the `uploads` table set to NULL.

File: backend/entityservice/encoding_storage.py (hunk 5 above)
Comment (hardbyte) [reply backend/entityservice/encoding_storage.py:158]: thanks, that looks like an oversight on my part. Clearly prequeued runs with external data are not covered by a test :-/

File: backend/entityservice/api_def/openapi.yaml [hunk 9]
```diff
@@ -1077,17 +1087,92 @@ components:
...
+      required: [encodings]
+      properties:
+        encodings:
+          oneOf:
+            - $ref: '#/components/schemas/EncodingArray'
+            - $ref: '#/components/schemas/ExternalData'
+        blocks:
+          oneOf:
+            - $ref: '#/components/schemas/BlockMap'
+            ## TODO may be useful to handle external blocking data too
```
[the remark this answers is not in the document]
Comment (hardbyte) [reply backend/entityservice/api_def/openapi.yaml:1102]: This will be part of a follow up PR

**Comment by hardbyte** [comment]
Changes incorporated in PR #551

**Pull request closed without merging**